tests/mock_crust/client_restrictions.rs

```diff
@@ -79,3 +79,42 @@ fn only_one_client_per_ip() {
     let _ = poll_all(&mut nodes, &mut clients);
     expect_next_event!(unwrap!(clients.last_mut()), Event::Terminate);
 }
+
+/// Reconnect a client (disconnected as network not having enough nodes) with the same id.
+#[test]
+fn reconnect_disconnected_client() {
+    let network = Network::new(MIN_SECTION_SIZE, None);
+    let mut nodes = create_connected_nodes(&network, MIN_SECTION_SIZE - 1);
+
+    let config = Some(BootstrapConfig::with_contacts(
+        &[nodes[1].handle.endpoint()],
+    ));
+    let full_id = FullId::new();
+
+    // Client will get rejected as network not having enough nodes.
+    let mut clients =
+        vec![
+            TestClient::new_with_full_id(&network, config.clone(), None, full_id.clone()),
+        ];
+    let _ = poll_all(&mut nodes, &mut clients);
+    expect_next_event!(unwrap!(clients.last_mut()), Event::Terminate);
+
+    let _ = clients.remove(0);
```

Comment:
Yeh might be better just re creating the client here to be safe, as we might have to re-init the state machine and other objects which might be in an incompatible state after firing the terminate event to re-use later on.